web/admin_service.go

```diff
@@ -0,0 +1,161 @@
+func (s *AdminService) CreateCourse(ctx context.Context, in *pb.Course) (*pb.Course, error) {
+	usr, scm, err := s.getUserAndSCM(ctx, in.GetID())
+	if err != nil {
+		s.logger.Errorf("CreateCourse failed: scm authentication error: %v", err)
+		return nil, ErrInvalidUserInfo
+	}
+	if !usr.IsAdmin {
+		s.logger.Error("CreateCourse failed: user is not admin")
+		return nil, status.Error(codes.PermissionDenied, "user must be admin to create course")
+	}
+
+	// make sure that the current user is set as course creator
+	in.CourseCreatorID = usr.GetID()
+	course, err := s.createCourse(ctx, scm, in)
+	if err != nil {
+		s.logger.Errorf("CreateCourse failed: %v", err)
+		// errors informing about requested organization state will have code 9: FailedPrecondition
+		// error message will be displayed to the user
+		if contextCanceled(ctx) {
+			return nil, status.Error(codes.FailedPrecondition, ErrContextCanceled)
+		}
+		if err == ErrAlreadyExists || err == ErrFreePlan {
+			return nil, status.Error(codes.FailedPrecondition, err.Error())
+		}
+		if ok, parsedErr := parseSCMError(err); ok {
+			return nil, parsedErr
+		}
+		return nil, status.Error(codes.InvalidArgument, "failed to create course")
+	}
+	return course, nil
+}
+
+// TODO(vera): these methods are temporary duplicate and will be removed,
+// all services will need to somehow fetch current user and the proper scm client.
+// The goal is to do it some other way. User info will be in
+// the JWT token, scm client will "belong" to a course, not user
+
+func (s *AdminService) getUserAndSCM(ctx context.Context, courseID uint64) (*pb.User, scm.SCM, error) {
+	usr, err := s.getCurrentUser(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+	scm, err := s.getSCM(courseID)
+	if err != nil {
+		return nil, nil, err
+	}
+	return usr, scm, nil
+}
+
+func (s *AdminService) getCurrentUser(ctx context.Context) (*pb.User, error) {
+	// process user id from context
+	meta, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return nil, errors.New("malformed request")
+	}
+	userValues := meta.Get("user")
+	if len(userValues) == 0 {
+		return nil, errors.New("no user metadata in context")
+	}
+	if len(userValues) != 1 || userValues[0] == "" {
+		return nil, errors.New("invalid user payload in context")
+	}
+	userID, err := strconv.ParseUint(userValues[0], 10, 64)
+	if err != nil {
+		return nil, err
+	}
+	// return the user corresponding to userID, or an error.
+	return s.db.GetUser(userID)
+}
+
+// TODO(vera): repurpose for new scm type (or two scm types)
+func (s *AdminService) getSCM(courseID uint64) (scm.SCM, error) {
+	sc, ok := s.app.GetSCM(courseID)
+	if ok {
+		return sc, nil
+	}
+	return nil, errors.New("no SCM found")
+}
+
+// TODO(vera): this method must work for any service, all services must share same scm list
+// MakeSCMClients creates a new scm client (GitHub app installation based client) for each course in the database.
```

Comment:
Should this be moved to the AdminService constructor? Since it should be called at start and shouldn't be called multiple times... if I understood correctly.